streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java

```diff
@@ -93,47 +95,15 @@ public TopicPartition partition() {
      */
     int addRawRecords(final Iterable<ConsumerRecord<byte[], byte[]>> rawRecords) {
         for (final ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
-
-            final ConsumerRecord<Object, Object> record = recordDeserializer.deserialize(processorContext, rawRecord);
-            if (record == null) {
-                // this only happens if the deserializer decides to skip. It has already logged the reason.
-                continue;
-            }
-
-            final long timestamp;
-            try {
-                timestamp = timestampExtractor.extract(record, timeTracker.get());
-            } catch (final StreamsException internalFatalExtractorException) {
-                throw internalFatalExtractorException;
-            } catch (final Exception fatalUserException) {
-                throw new StreamsException(
-                    String.format("Fatal user code error in TimestampExtractor callback for record %s.", record),
-                    fatalUserException);
-            }
-            log.trace("Source node {} extracted timestamp {} for record {}", source.name(), timestamp, record);
-
-            // drop message if TS is invalid, i.e., negative
-            if (timestamp < 0) {
-                log.warn(
-                    "Skipping record due to negative extracted timestamp. topic=[{}] partition=[{}] offset=[{}] extractedTimestamp=[{}] extractor=[{}]",
-                    record.topic(), record.partition(), record.offset(), timestamp, timestampExtractor.getClass().getCanonicalName()
-                );
-                ((StreamsMetricsImpl) processorContext.metrics()).skippedRecordsSensor().record();
-                continue;
-            }
-
-            final StampedRecord stampedRecord = new StampedRecord(record, timestamp);
-            fifoQueue.addLast(stampedRecord);
-            timeTracker.addElement(stampedRecord);
+            fifoQueue.addLast(rawRecord);
         }
 
-        // update the partition timestamp if its currently
-        // tracked min timestamp has exceed its value; this will
-        // usually only take effect for the first added batch
-        final long timestamp = timeTracker.get();
+        maybeUpdateHeadRecord();
 
-        if (timestamp > partitionTime) {
-            partitionTime = timestamp;
+        // update the partition timestamp if the current head
+        // record's timestamp has exceed its value
+        if (headRecord != null && headRecord.timestamp > partitionTime) {
+            partitionTime = headRecord.timestamp;
```

Comment:
That's a good question :) Originally I changed it so that the partition st would not be advanced after a record is popped out immediately, but then I realized we cannot do that since we rely on partition st to determine which partition's queue to pop from. So in the latest PR I've reverted this back.

The logic is a bit different from your suggestions, the logic is that after we've added to the raw record queue, we check if the head record needs to be updated (if it was null, and queue is not empty), and we only update partition ts when the record was updated.